flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java

```diff
@@ -196,183 +208,346 @@ public synchronized void start() {
+      LOG.debug("Exception follows.", ex);
+      // We don't propogate the exception as the transaction would have been
+      // rolled back and we can still finish stopping
     }
 
-    // signal that this sink has stopped
+  // signal that this sink has stopped
     LOG.info("Stopped dataset sink: " + getName());
     super.stop();
   }
 
   @Override
   public Status process() throws EventDeliveryException {
-    if (writer == null) {
-      try {
-        this.writer = newWriter(login, target);
-      } catch (DatasetException e) {
-        // DatasetException includes DatasetNotFoundException
-        throw new EventDeliveryException(
-          "Cannot write to " + getName(), e);
+    long processedEvents = 0;
+
+    try {
+      if (shouldRoll()) {
+        closeWriter();
+        commitTransaction();
+        createWriter();
       }
-    }
 
-    // handle file rolling
-    if ((System.currentTimeMillis() - lastRolledMs) / 1000 > rollIntervalS) {
-      // close the current writer and get a new one
-      writer.close();
-      this.writer = newWriter(login, target);
-      this.lastRolledMs = System.currentTimeMillis();
-      LOG.info("Rolled writer for " + getName());
-    }
+      // The writer shouldn't be null at this point
+      Preconditions.checkNotNull(writer,
+          "Can't process events with a null writer. This is likely a bug.");
+      Channel channel = getChannel();
 
-    Channel channel = getChannel();
-    Transaction transaction = null;
-    try {
-      long processedEvents = 0;
+      // Enter the transaction boundary if we haven't already
+      enterTransaction(channel);
 
-      transaction = channel.getTransaction();
-      transaction.begin();
       for (; processedEvents < batchSize; processedEvents += 1) {
         Event event = channel.take();
+
         if (event == null) {
           // no events available in the channel
           break;
         }
 
-        this.datum = deserialize(event, reuseDatum ? datum : null);
-
-        // writeEncoded would be an optimization in some cases, but HBase
-        // will not support it and partitioned Datasets need to get partition
-        // info from the entity Object. We may be able to avoid the
-        // serialization round-trip otherwise.
-        writer.write(datum);
+        write(event);
       }
 
-      // TODO: Add option to sync, depends on CDK-203
-      writer.flush();
-
-      // commit after data has been written and flushed
-      transaction.commit();
-
-      if (processedEvents == 0) {
-        counter.incrementBatchEmptyCount();
-        return Status.BACKOFF;
-      } else if (processedEvents < batchSize) {
-        counter.incrementBatchUnderflowCount();
-      } else {
-        counter.incrementBatchCompleteCount();
+      // commit transaction
+      if (commitOnBatch) {
+        // Sync before commiting. A failure here will result in rolling back
+        // the transaction
+        writer.sync();
```

Comment:
I can note the Flushable/Syncable interface check to the list of follow-up
improvements that require a new kite release.

Do you thank an option like kite.durabilityMethod with values of sync or
flush makes sense?
On Tue, Jan 27, 2015 at 08:32 Tom White <notifications@github.com> wrote:

> In
> flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java
> <https://github.com/joey/flume/pull/1#discussion_r23620845>:
>
> > -
> > -      // commit after data has been written and flushed
> > -      transaction.commit();
> > -
> > -      if (processedEvents == 0) {
> > -        counter.incrementBatchEmptyCount();
> > -        return Status.BACKOFF;
> > -      } else if (processedEvents < batchSize) {
> > -        counter.incrementBatchUnderflowCount();
> > -      } else {
> > -        counter.incrementBatchCompleteCount();
> > +      // commit transaction
> > +      if (commitOnBatch) {
> > +        // Sync before commiting. A failure here will result in rolling back
> > +        // the transaction
> > +        writer.sync();
>
> It might be worth having an option to flush rather than sync. Sync is more
> expensive, and also isn't supported by some dataset implementations (e.g.
> HBase).
>
> Also, this should be updated post
> https://issues.cloudera.org/browse/CDK-863 to test to see if the writer
> is Flushable or Syncable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joey/flume/pull/1/files#r23620845>.
>